adapter: connection validation on creation and on demand

petrosagg opened the pull request:
### Motivation

This PR adds a new feature giving the ability to users to validate `CONNECTION` objects on creation and on demand. This accomplished by two modifications in the SQL we accept:

**Validation on creation:** All `CREATE CONNECTION` statements can now be modified with the familiar "WITH" options. For now the only option that can be specified is the `VALIDATE` option whose value must be either `true` or `false`. When the option is not specified a sensible default is picked depending on the type of connection. The default values can be seen in the following table:

| Type            | Validate by default |
|-----------------|---------------------|
| Postgres        | Yes                 |
| Kafka           | Yes                 |
| Schema Registry | Yes                 |
| AWS             | No                  |
| AWS Privatelink | No                  |
| SSH             | No                  |

**Validation on demand:** All `CONNECTION` objects can be validated on demand using a new `VALIDATE CONNECTION <name>` statement. This query will block until the validation completes and will either successfully finish, returning no rows, or it will return SQL error with details about what went wrong.

**NOTE:** This PR does *NOT* actually implement validation for `SSH`, `AWS`, and `AWS PRIVATELINK` connections. This will be done in a separate PR to leave this one focused on the validation infrastructure. `KAFKA`, `POSTGRES`, and `CSR` connections are validated as part of the PR since it has trivial to do. Of course if a `KAFKA`/`POSTGRES`/`CSR` connection *does* use ssh/privatelink then those are implicitly validated too.

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

sploiselle commented on src/sql/src/pure.rs:
```diff
@@ -853,6 +855,34 @@ async fn purify_source_format_single(
+            let config = conn.config(&*connection_context.secrets_reader).await?;
+            config.connect("connection validation").await?;
+        }
+        Connection::Aws(_conn) => {}
```
Do these error somewhere else? I'd expect them to be unreachable or error.

sploiselle commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -554,11 +557,34 @@ impl<T: AstInfo> AstDisplay for CreateConnectionStatement<T> {
         f.write_str(" TO ");
-        self.connection.fmt(f)
+        self.connection.fmt(f);
+        if let Some(validate) = self.validate {
```
I am very leery of pretending to use the `WITH` options structure without actually using the same machinery as everywhere else. I realize it's a little annoying  for a single option, but believe the consistency and hygiene is worth the extra bit of effort.

petrosagg replied:
I went through the same dilemma and convinced myself that it's ok but since it smells to you too I'll do it the proper way

mjibson commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -430,6 +433,47 @@ impl Coordinator {
+        // Ensure that all dependencies still exist after validation, as a
+        // `DROP SECRET` may have sneaked in. If any have gone missing, we
+        // revalidate the original statement. This will either produce a nice
+        // "unknown secret" error, or pick up a new connector that has
```
How does this pick up a new connector? Is the ID reused? We document elsewhere we assume they never are: https://github.com/MaterializeInc/materialize/blob/149d2e5bca1f54e94acfe02ab5f5751b9cb2adfc/src/adapter/src/coord.rs#L399

petrosagg replied:
This comment is referring to any object dependencies a `CREATE CONNECTION` statement might have and not the connection being created. If any of the dependencies disappear between the initial connection planning and sequencing, which might happen because validation takes a while, then we will replan the statement and try again. During replanning the object names referred by the original `CREATE CONNECTION ..` statement will either resolve to new objects or produce "uknown object X" errors

mjibson commented on src/adapter/src/coord/message_handler.rs:
```diff
@@ -430,6 +433,47 @@ impl Coordinator {
+    #[tracing::instrument(level = "debug", skip(self, tx, session))]
+    async fn message_create_connection_validation_ready(
+        &mut self,
+        CreateConnectionValidationReady {
+            session,
+            tx,
+            result,
+            params,
+            original_stmt,
+            depends_on,
+            otel_ctx,
+        }: CreateConnectionValidationReady,
+    ) {
+        otel_ctx.attach_as_parent();
+
+        // Ensure that all dependencies still exist after validation, as a
+        // `DROP SECRET` may have sneaked in. If any have gone missing, we
+        // revalidate the original statement. This will either produce a nice
+        // "unknown secret" error, or pick up a new connector that has
+        // replaced the dropped connector.
+        //
+        // WARNING: If we support `ALTER SECRET`, we'll need to also check
+        // for connectors that were altered while we were purifying.
+        if !depends_on
+            .iter()
+            .all(|id| self.catalog().try_get_entry(id).is_some())
+        {
+            self.handle_execute_inner(original_stmt, params, session, tx)
+                .await;
+            return;
+        }
+
+        let plan = match result {
+            Ok(ok) => ok,
+            Err(e) => return tx.send(Err(e), session),
+        };
+
+        self.sequence_plan(tx, session, Plan::CreateConnection(plan), depends_on)
+            .await;
+    }
```
The `PeekValidity` struct does have some logic to handle things like this. If it could be reused and renamed to be more general, I'm on board with that.

petrosagg replied:
This is following the source purification pattern right above it. What is the issue that the other PR had?

def- commented on src/sql/src/plan/statement/validate.rs:
```diff
@@ -0,0 +1,43 @@
+    Ok(StatementDesc::new(None))
+}
+
+pub fn plan_validate_connection(
```
This should be covered by tests, right? Strangely code coverage is not showing it, I guess a bug in coverage.

petrosagg replied:
It's not a bug in coverage, I will add some tests for the `VALIDATE CONNECTION` statements

def- reviewed (commented):
Nightly looks fine: https://buildkite.com/materialize/nightlies/builds/2626
Coverage: https://buildkite.com/materialize/coverage/builds/145#_

benesch commented:
Very nice! We should get @MaterializeInc/devex on testing this, too (cc @morsapaes). @sjwiesman is also quite excited about this feature.

In terms of rollout: are we comfortable with this being insta-stable? Would we prefer to put this behind a LD flag first and get some experience with it in staging/canaries first?

petrosagg commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -549,6 +553,53 @@ impl Coordinator {
+                    }
+                };
+
+                if plan.validate {
+                    let internal_cmd_tx = self.internal_cmd_tx.clone();
+                    let conn_id = session.conn_id().clone();
+                    let connection_context = self.connection_context.clone();
+                    let otel_ctx = OpenTelemetryContext::obtain();
+                    task::spawn(|| format!("validate_connection:{conn_id}"), async move {
+                        let connection = &plan.connection.connection;
+                        let result = match connection.validate(&connection_context).await {
+                            Ok(()) => Ok(plan),
+                            Err(err) => Err(err.into()),
+                        };
+
+                        // It is not an error for validation to complete after `internal_cmd_rx` is dropped.
+                        let result =
+                            internal_cmd_tx.send(Message::CreateConnectionValidationReady(
+                                CreateConnectionValidationReady {
+                                    session,
+                                    tx,
+                                    result,
+                                    params,
+                                    depends_on,
+                                    original_stmt,
+                                    otel_ctx,
+                                },
+                            ));
+                        if let Err(e) = result {
+                            tracing::warn!("internal_cmd_rx dropped before we could send: {:?}", e);
+                        }
+                    });
```
I don't know if it makes sense to go in the `sequence` stage because this operation is not sequenced until validation happens and other things might get sequenced in between. Additionally, in the event that we cannot actually sequence a `CREATE CONNECTION` after its validation succeeds (it can happen if in the meantime you drop one of its dependencies) then we need access to the original statement so that we can re-plan it. Arguably we should have a mechanism to lock all those in-flight resources from being dropped but we don't have that now

jkosh44 replied:
I'm going to delegate this part of the review to @mjibson. He's been pulling things off of the main coord thread, so he'd better understand what belongs where.

jkosh44 replied:
Bumping this comment (@mjibson) I just want to make sure this is consistent with the work you did to move things off the main coord thread.

petrosagg commented on src/storage-client/src/types/connections.rs:
```diff
@@ -165,6 +168,35 @@ pub enum Connection {
+    /// Whether this connection should be validated by default on creation.
+    pub fn validate_by_default(&self) -> bool {
+        match self {
+            Connection::Kafka(conn) => conn.validate_by_default(),
+            Connection::Csr(conn) => conn.validate_by_default(),
+            Connection::Postgres(conn) => conn.validate_by_default(),
+            Connection::Ssh(conn) => conn.validate_by_default(),
+            Connection::Aws(conn) => conn.validate_by_default(),
+            Connection::AwsPrivatelink(conn) => conn.validate_by_default(),
```
Some connections cannot be validated on creation because there is additional setup steps to be done that happen after the connection is created. For example creating an SSH connection requires getting the mz public keys and putting them in your bastion but you only get access to them once the connection is created. At the same time we want to have validation by default for the connections that are expected to work in the common cases.

sjwiesman commented:
> In terms of rollout: are we comfortable with this being insta-stable? Would we prefer to put this behind a LD flag first and get some experience with it in staging/canaries first?

I would like my hands on this as soon as possible. If there are stability concerns, what about setting validate by default to `false` for all connections and then switching back as we grow more comfortable?

benesch commented:
> > In terms of rollout: are we comfortable with this being insta-stable? Would we prefer to put this behind a LD flag first and get some experience with it in staging/canaries first?
> 
> I would like my hands on this as soon as possible. If there are stability concerns, what about setting validate by default to `false` for all connections and then switching back as we grow more comfortable?

Oh, that wasn't the stability concern I had, but that's a good idea. @petrosagg we should definitely add a LD flag like `enable_default_connection_validation` that we can flip to false in an emergency to unblock a customer.

The stability concerns I have are around the syntax. We're committing to support `CREATE CONNECTION ... WITH (VALIDATE = {TRUE | FALSE})` and `VALIDATE CONNECTION <name>` approximately forever with this PR. It's a very small amount of new syntax surface area, though, so I'm comfortable with it if everyone else is.

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -3494,14 +3495,40 @@ pub fn plan_create_connection(
     };
     let name = scx.allocate_qualified_name(normalize::unresolved_item_name(name)?)?;
 
+    let mut user_validate = None;
+    for option in with_options {
+        match option.name {
+            CreateConnectionOptionName::Validate => {
+                scx.require_feature_flag(&vars::ENABLE_CONNECTION_VALIDATION_SYNTAX)?;
+                if user_validate.is_some() {
+                    sql_bail!("Option VALIDATE must be set at most once");
+                }
+                user_validate = match option.value {
+                    Some(WithOptionValue::Value(Value::Boolean(new_val))) => Some(new_val),
+                    None => Some(true),
+                    _ => sql_bail!("Option VALIDATE must be set to either TRUE or FALSE"),
+                };
+            }
+        }
+    }
```
We have a macro that handles this sort of validation now! Grep around for `generate_extracted_config!`.

petrosagg replied:
Oh cool! Fixed

petrosagg commented on src/storage-client/src/types/connections.rs:
```diff
@@ -1036,3 +1117,24 @@ impl SshTunnel {
+    #[allow(clippy::unused_async)]
+    async fn validate(&self, _connection_context: &ConnectionContext) -> Result<(), anyhow::Error> {
+        Ok(())
+    }
```
Added an error saying that validating is not supported yet

petrosagg commented:
@benesch I added the syntax behind a feature flag and the testing by default behind a system variable that is set to `true` by default. We can disable it on an environment by environment basis if we see issues with validation.

@jkosh44 I think I fixed your RBAC comments, can you check the commit "fix rbac permissions" commit? I separated out the fixes there.

@def- I added tests for `VALIDATE CONNECTION` there should be coverage now.

I think this should be good to merge now

jkosh44 commented on src/adapter/src/rbac.rs:
```diff
@@ -641,6 +644,7 @@ fn generate_required_ownership(plan: &Plan) -> Vec<ObjectId> {
         Plan::AlterSource(plan) => vec![ObjectId::Item(plan.id)],
         Plan::AlterItemRename(plan) => vec![ObjectId::Item(plan.id)],
         Plan::AlterSecret(plan) => vec![ObjectId::Item(plan.id)],
+        Plan::ValidateConnection(plan) => vec![ObjectId::Item(plan.id)],
```
I don't think this is correct, unless you want to restrict validating a connection to the owner of that connection.

benesch commented on src/storage-client/src/types/connections.rs:
```diff
@@ -372,6 +404,36 @@ impl KafkaConnection {
+            .create_with_context(connection_context, MzClientContext, &BTreeMap::new())
+            .await?;
+
+        // librdkafka doesn't expose an API for determining whether a connection to
+        // the Kafka cluster has been successfully established. So we make a
+        // metadata request, though we don't care about the results, so that we can
+        // report any errors making that request. If the request succeeds, we know
+        // we were able to contact at least one broker, and that's a good proxy for
+        // being able to contact all the brokers in the cluster.
+        //
+        // The downside of this approach is it produces a generic error message like
+        // "metadata fetch error" with no additional details. The real networking
+        // error is buried in the librdkafka logs, which are not visible to users.
+        //
+        // TODO(benesch): pull out more error details from the librdkafka logs and
+        // include them in the error message.
```
Can we structure the code to avoid copy/pasting this whole comment? Maybe a shared `validate` method in the `kafka-util` crate?

petrosagg replied:
Oops, I had already done this and forgot to delete kafka-util version which was dead code. Deleted!

benesch reviewed (commented):
I only re-reviewed the `generate_extracted_config!` bit, which LGTM. Still deferring the rest of review to other folks!

benesch commented:
> @benesch I added the syntax behind a feature flag and the testing by default behind a system variable that is set to `true` by default. We can disable it on an environment by environment basis if we see issues with validation.

Looks great, thank you!

jkosh44 reviewed (approved):
LGTM
